### PR TITLE
enable both dicts and list/tuple batches in trainers

### PR DESCRIPTION
**Pull request by petermartigny**
In Transfer NLP we introduced the `DataFrameDataset` that enables pytorch dataloaders to throw map-like batches. This enabled flexibility in the numbe rof input we feed to the model's `forward` methods and the number of pred we would want to test on.

A drawback is that a user has to make sure that the keys the map-like batches match the exact names of inputs of the `forward` methods.

In this PR, we add the possiblity to feed batches which are lists or tuples, and for those cases we feed `*batch[:-1]` to the model's `forward` and we use `batch[-1]` as target (instead of the hard-coded `batch["y_target"]`).

### CODE REVIEW

File: transfer_nlp/plugins/trainers.py [hunk 1]
```diff
@@ -224,10 +233,11 @@ def log_training_validation_results(trainer):
 
         @self.trainer.on(Events.COMPLETED)
         def log_test_results(trainer):
-            self.evaluator.run(self.dataset_splits.test_data_loader())
-            metrics = self.evaluator.state.metrics
-            store_metrics(metrics=metrics, mode="test")
-            logger.info(f"Test Results - Epoch: {trainer.state.epoch} {print_metrics(metrics)}")
+            if self.dataset_splits.test_set:
```
Comment (petermartigny) [inline transfer_nlp/plugins/trainers.py]: If we don't provide a test set when experimenting we don't want this block

File: transfer_nlp/plugins/trainers.py [hunk 2]
```diff
@@ -380,8 +390,14 @@ def update_engine(self, engine, batch):
 
         self.model.train()
         batch = _prepare_batch(batch, device=self.device, non_blocking=False)
-        y_pred = self._forward(batch)
-        loss = self.loss(input=y_pred, target=batch['y_target'])
+        if isinstance(batch, dict):
```
Comment (petermartigny) [inline transfer_nlp/plugins/trainers.py]: Main change of the PR is here

File: transfer_nlp/plugins/trainers.py [hunk 3]
```diff
@@ -395,16 +411,26 @@ def update_engine(self, engine, batch):
...
         self.model.eval()
         with torch.no_grad():
             batch = _prepare_batch(batch, device=self.device, non_blocking=False)
-            y_pred = self._forward(batch)
-            return self.eval_output_transform(y_pred, batch['y_target'])
+            if isinstance(batch, dict):
+                y_pred = self._forward(batch)
+                return self.eval_output_transform(y_pred, batch['y_target'])
+            elif isinstance(batch, tuple) or isinstance(batch, list):
+                y_pred = self.model.forward(*batch[:-1])
```
Comment (kireet) [inline transfer_nlp/plugins/trainers.py]: is this right? when doing inference you may not have labels.

Comment (kireet) [reply transfer_nlp/plugins/trainers.py]: i guess the old dict approach also required y_target...

Comment (petermartigny) [reply transfer_nlp/plugins/trainers.py]: The idea here is to use ignite trainers for training only (which is, ignite only sees labeled data in training and evaluation/test time), and then use model's calls for unlabeled inference

File: transfer_nlp/plugins/trainers.py [hunk 4]
```diff
@@ -210,7 +219,7 @@ def log_training_validation_results(trainer):
             self.evaluator.run(self.dataset_splits.train_data_loader())
             metrics = self.evaluator.state.metrics
             store_metrics(metrics=metrics, mode="training")
-            # logger.info(f"Training Results - Epoch: {trainer.state.epoch} {print_metrics(metrics)}")
+            logger.info(f"Training Results - Epoch: {trainer.state.epoch} {print_metrics(metrics)}")
```
[the remark this answers is not in the document]
Comment (petermartigny) [reply transfer_nlp/plugins/trainers.py]: not really, we already use it for validation metrics
